### ISSUE TITLE
@babel/plugin-transform-classses(createSuper)@7.9.x may have performance issue

### ISSUE THREAD

**Issue by shrinktofit** [babel/babel#11356]
Classes **with super class** are very slow to `new`, in comparison to what babel@7.8.8 does. The difference can be found at [here](https://fossies.org/diffs/babel/7.8.8_vs_7.9.0/packages/babel-plugin-transform-classes/test/fixtures/regression/T2997/output.js-diff.html).

```js
function test(Vec3) {
    const t1 = Date.now();
    (() => {
        for (let i = 0; i < 1000000; ++i) {
            const x = Math.random();
            new Vec3(x, x, x); // Key line
        }
    })();
    const t2 = Date.now();
    console.log(`Past: ${t2 - t1}ms`);
}

test(require('./out/7.8.8/vec3.js').Vec3); // 18ms
test(require('./out/7.9.x/vec3.js').Vec3); // Oops, 2451ms!!
```

I have upload the test code to [this repository](https://github.com/shrinktofit/babel-perf). You can try it just by `node ./run-perf.js`.

**Comment by jridgewell** [comment]
Unfortunately, this is the cost of `Reflect.construct` with a `NewTarget`. We could better optimize `_createSuper` to avoid duplicate code checks and return specialized wrappers based on environment, but we're still going to be much slower.

**Comment by nicolo-ribaudo** [comment]
We can at least disable `createSuper` (or use a reduced version) in loose mode.

In "normal mode" the first goal is to be spec-compliant and fully compatible, but in loose mode we can assume that everything is used "normally" and prefer speed over correctness.

**Comment by jridgewell** [comment]
> We can at least disable createSuper (or use a reduced version) in loose mode.

Oh, I thought we avoided that for loose already. We should definitely do that.

**Comment by nicolo-ribaudo** [comment]
@shrinktofit I investigated a bit the performance issue, here is what I found:
- The performance regression only happens in engines with native `Reflect.construct` support. You can test it, for example, in Node.js 5 (I had to replace `let`/`const` with `var`, and to remove destructuring).
- All the engines I know with `Reflect.construct` also support native classes, so maybe you can avoid transpiling them for those engines.
- When using `loose` mode, there isn't any performance difference between 7.8.8 and 7.9.x
- If in the `createSuper` helper we move the call to `_isNativeReflectConstruct` in the outer function, we get a ~5% speed improvement.

**Comment by shrinktofit** [comment]
> When using loose mode, there isn't any performance difference between 7.8.8 and 7.9.x

This looks like the most appropriate solution for us, for now, thank you~

@nicolo-ribaudo  BTW, why can't we substitute `Reflect.construct` with operator new?

**Comment by nicolo-ribaudo** [comment]
It's because with `new` you can't specify the third parameter of `Reflect.construct` (`new.target`).

**Comment by benmccann** [comment]
Is there anyway to turn off just this feature without enabling `loose`? It's causing a 10x slowdown on our library overall, but our code breaks with `loose`

**Comment by jridgewell** [comment]
You could delete `Reflect.construct`, which should bring it back to almost 7.8's speed.

Or, we may be able to add an explicit option to the transorm?

**Comment by nicolo-ribaudo** [comment]
I'd be ok with a `assumeNoExtendNative` option to the classes plugin.

**Comment by shrinktofit** [comment]
> I'd be ok with a `assumeNoExtendNative` option to the classes plugin.
@nicolo-ribaudo  Any progress on this option?

**Comment by JLHwung** [comment]
On the `assumeNoExtendNative`, we can consider add it to the `"assumptions"` introduced in https://github.com/babel/babel/pull/12219

**Comment by nicolo-ribaudo** [comment]
I want to keep that RFC/PR without adding new capabilities to the transforms (so that it only focuses on moving the assumptions from plugins to the top level), but yes this is a good candidate for a new assumption.

**Comment by shrinktofit** [comment]
@nicolo-ribaudo  Will this be added into the big list of "assumptions"?

**Comment by nicolo-ribaudo** [comment]
Now that we shipped `assumptions` in 7.13.0, I was looking into this. Does the [`superIsCallableConstructor`](https://babeljs.io/assumptions#superIsCallableConstructor) assumption work for you? It basically skips the `createSuper` call.

**Comment by nicolo-ribaudo** [comment]
I'm closing this since _I think_ `superIsCallableConstructor` does exactly what's needed for this feature request. If I'm wrong ping me and I'll reopen this issue.
